lxd/db/cluster/update.go

```diff
@@ -70,6 +70,241 @@ var updates = map[int]schema.Update{
 	31: updateFromV30,
 	32: updateFromV31,
 	33: updateFromV32,
+	34: updateFromV33,
+}
+
+// Remove multiple entries of the same volume when using remote storage.
+// Also, allow node ID to be null for the instances and storage_volumes tables, and set it to null
+// for instances and storage volumes using remote storage.
+func updateFromV33(tx *sql.Tx) error {
+	stmts := `
+SELECT storage_volumes.id, storage_volumes.name
+FROM storage_volumes
+JOIN storage_pools ON storage_pools.id=storage_volumes.storage_pool_id
+WHERE storage_pools.driver IN ("ceph", "cephfs")
+ORDER BY storage_volumes.name
+`
+
+	// Get the total number of storage volume rows.
+	count, err := query.Count(tx, "storage_volumes", "")
+	if err != nil {
+		return errors.Wrap(err, "Failed to get storage volumes count")
+	}
+
+	volumes := make([]struct {
+		ID   int
+		Name string
+	}, count)
+	dest := func(i int) []interface{} {
+		return []interface{}{
+			&volumes[i].ID,
+			&volumes[i].Name,
+		}
+	}
+
+	stmt, err := tx.Prepare(stmts)
+	if err != nil {
+		return errors.Wrap(err, "Failed to prepary storage volume query")
+	}
+
+	err = query.SelectObjects(stmt, dest)
+	if err != nil {
+		return errors.Wrap(err, "Failed to fetch storage volumes")
+	}
+
+	// Remove multiple entries of the same volume when using remote storage
+	for i := 1; i < count; i++ {
```

Comment:
I believe here `count` might be greater than `len(volumes)`, hence leading to a crash. You should either modify `query.Count(tx, "storage_volumes", "")` to include the same filters as the `query.SelectObjects` call, or make your `dest` function grow the volumes slice each time its called.

Reply:
It won't crash as the size of `volumes` is set to `count`.